### ISSUE TITLE
Type aliases inside dataclass definitions are not supported at runtime inside of if TYPE_CHECKING

### ISSUE THREAD

**Issue by eltoder** [python/mypy#19706]
**Bug Report**

Mypy produces a "Type aliases inside dataclass definitions are not supported at runtime" even for code inside of `if TYPE_CHECKING:` blocks, which are not going to be executed at runtime. This error is correct outside of `if TYPE_CHECKING:`. Furthermore, wrapping `TypeAlias` in a `ClassVar` to avoid the error satisfies mypy, but the resulting type alias becomes `Any`.

**To Reproduce**

```python
from dataclasses import dataclass
from typing import TYPE_CHECKING, ClassVar, TypeAlias, assert_type

if TYPE_CHECKING:
    @dataclass
    class DC:
        T: TypeAlias = int

@dataclass
class DC2:
    T: ClassVar[TypeAlias] = int

def test(x1: DC.T, x2: DC2.T) -> None:
    assert_type(x1, int)
    assert_type(x2, int)
```
https://mypy-play.net/?mypy=latest&python=3.12&gist=e0e7315cc293f94cbb480ade20488afe

**Expected Behavior**

At least one of these should work without errors. Currently, DC generates a mypy error, but produces a working type alias. DC2 does not generate any errors, but DC2.T is `Any` instead of `int`.

**Actual Behavior**

```
main.py:7: error: Type aliases inside dataclass definitions are not supported at runtime  [misc]
main.py:15: error: Expression is of type "Any", not "int"  [assert-type]
```

**Your Environment**

- Mypy version used: 1.17.1
- Python version used: 3.12

**Comment by tmke8** [comment]
I started a discussion about this a few years ago, but there wasn't enough interest in it: https://github.com/python/typing/discussions/1124

I also opened an issue at CPython: https://github.com/python/cpython/issues/91322

One thing that has changed since then is the new type aliases from Python 3.12. This works:

```python
@dataclass
class DC2:
    type T = int
```

and also this:
```python
from typing_extensions import TypeAliasType
@dataclass
class DC2:
    T = TypeAliasType("T", int)
```
